### ISSUE TITLE
Hap: ResolutionAlmostCrystalQuotient produces error

### ISSUE THREAD

**Issue by dominicelse** [gap-system/gap#1931]
Running the code

```
LoadPackage("Hap");
LoadPackage("Cryst");
G := SpaceGroupIT(3,207);
Gp := Image(IsomorphismPcpGroup(G));
R := ResolutionAlmostCrystalQuotient(Gp,4,1);
```

### Observed behaviour
```
Error, gens and imgs must have same length called from
GroupGeneralMappingByImagesNC( G, H, gens, imgs 
 ) at /home/dom/source/gap/gap4r8/lib/ghom.gi:443 called from
GroupHomomorphismByImagesNC( GD, G, GeneratorsOfGroup( GD ), GeneratorsOfPcp( pcpGD ) 
 ) at /home/dom/source/gap/gap4r8/pkg/Hap1.11/lib/Resolutions/resACquotient.gi:58 called from
<function "ResolutionAlmostCrystalQuotient">( <arguments> )
 called from read-eval loop at line 5 of bug.gi
```

### Copy and paste GAP banner (to tell us about your setup)

```
 ┌───────┐   GAP 4.8.7, 24-Mar-2017, build of 2017-06-15 19:42:38 (PDT)
 │  GAP  │   https://www.gap-system.org
 └───────┘   Architecture: x86_64-pc-linux-gnu-gcc-default64
 Libs used:  gmp, readline
 Loading the library and packages ...
 Components: trans 1.0, prim 2.1, small* 1.0, id* 1.0
 Packages:   AClib 1.2, Alnuth 3.0.0, AtlasRep 1.5.1, AutPGrp 1.8, CRISP 1.4.4, Cryst 4.1.12, 
             CrystCat 1.1.6, CTblLib 1.2.2, FactInt 1.5.4, FGA 1.3.1, GAPDoc 1.5.1, IRREDSOL 1.3.1, 
             LAGUNA 3.7.0, Polenta 1.3.7, Polycyclic 2.11, RadiRoot 2.7, ResClasses 4.6.0, Sophus 1.23, 
             SpinSym 1.5, TomLib 1.2.6, Utils 0.46
```

**Comment by alex-konovalov** [comment]
@dominicelse thanks - I've notified the HAP author.

It seems to me that the usage of `GroupHomomorphismByImagesNC` in HAP results in an invalid homomorphism object (which then triggers an error message from `GroupGeneralMappingByImages_for_pcp` function in the Polycyclic package).

```
brk> GeneratorsOfGroup(G);
[ g1, g2, g3, g4, id, id, id ]
brk> GeneratorsOfGroup(H);
[ g1, g2, g3, g4, g5, g6, g7 ]
brk> gens;
[ g1, g2, g3, g4, id, id, id ]
brk> imgs;
[ g1, g2, g3, g4 ]
```

**Comment by fingolfin** [comment]
The problem starts with these two lines in HAP:
```gap
GDhomG:=GroupHomomorphismByImagesNC(GD,G,    
GeneratorsOfGroup(GD),GeneratorsOfPcp(pcpGD));
```
Even without the length mismatch Alexander already diagnosed, this is incorrect, as it relies on the order of generators matching, which is true right now, but is *not* guaranteed by the specification of the functions involved in computing GD and pcpGD resp. their generating sets. However, `GDhomG` is actually never used! So, the first step of the fix is to simply remove the two offending lines.

Then we get a new error a few lines later, when computing `GDhomP`. This can be fixed by not using `Pcp` and instead mapping the generators of G into the groups P and GD. The patch below fixes the error for me.

```patch
diff --git a/Hap1.11/lib/Resolutions/resACquotient.gi b/Hap1.11/lib/Resolutions/resACquotient.gi
index d8c923b0..e03c05af 100755
--- a/Hap1.11/lib/Resolutions/resACquotient.gi
+++ b/Hap1.11/lib/Resolutions/resACquotient.gi
@@ -6,7 +6,7 @@ function(arg)
 local
 	G,K,KK,bool,
 	GhomP,P,T,Derived,i, RGD,
-	pcpGD, GD, GDhomPCGD,PCGD,GhomGD, GDhomG, GDhomP, TD, gensP, RP,RTD,
+	GD, GDhomPCGD,PCGD,GhomGD, GDhomP, TD, gensGD, gensP, RP,RTD,
         iso, isoTD;
 
 G:=arg[1];
@@ -29,7 +29,6 @@ for i in [2..KK] do
 Derived:=CommutatorSubgroup(G,Derived);
 od;
 
-pcpGD:=Pcp(G,Derived);
 GhomGD:=NaturalHomomorphism(G,Derived);
 GD:=Image(GhomGD);
 
@@ -52,12 +51,9 @@ if IsNilpotent(GD) and bool then
 
 else
 
-GDhomG:=GroupHomomorphismByImagesNC(GD,G,    
-GeneratorsOfGroup(GD),GeneratorsOfPcp(pcpGD));
-
-gensP:=List(GeneratorsOfPcp(pcpGD),x->Image(GhomP,x));
-GDhomP:=GroupHomomorphismByImagesNC(GD,P,          
-GeneratorsOfGroup(GD),gensP);
+gensGD:=List(GeneratorsOfGroup(G),x->Image(GhomGD,x));
+gensP:=List(GeneratorsOfGroup(G),x->Image(GhomP,x));
+GDhomP:=GroupHomomorphismByImagesNC(GD,P,gensGD,gensP);
 
 TD:=Kernel(GDhomP);
 
```

**Comment by alex-konovalov** [comment]
This now works with HAP 1.12.5 which is included in GAP 4.8.10 distribution:
```
gap> G := SpaceGroupIT(3,207);
SpaceGroupOnRightIT(3,207,'1')
gap> Gp := Image(IsomorphismPcpGroup(G));
Pcp-group with orders [ 2, 3, 2, 2, 0, 0, 0 ]
gap> R := ResolutionAlmostCrystalQuotient(Gp,4,1);
Resolution of length 4 in characteristic 0 for Pcp-group with orders 
[ 2, 3, 2, 2 ] . 

gap> 
```
